src/browser/awesomebar.js

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define((require, exports, module) => {
+
+  'use strict';
+
+  const MAX_RESULTS = 6;
+
+  const {DOM} = require('react')
+  const {fromJS} = require('immutable');
+  const Component = require('omniscient');
+  const {Element, Event} = require('./element');
+
+  const Awesomebar = Component('Awesomebar', (
+    {suggestionsCursor, theme},
+    {onOpen}) => {
+
+    let suggestions = [];
+
+    let search = suggestionsCursor.get('search').toArray();
+    let history = suggestionsCursor.get('history').toArray();
+
+    // searchCount + historyCount == 6
+    let searchCount = Math.min(search.length,
+                               MAX_RESULTS - Math.min(MAX_RESULTS / 2, history.length));
+    let historyCount = Math.min(history.length,
+                                MAX_RESULTS - Math.min(MAX_RESULTS / 2, search.length));
+
+    for (let i = 0; i < searchCount; i++) {
+      let href = search[i].get('href');
+      let text = search[i].get('text');
+      suggestions.push(DOM.p({
+        className: 'suggestion search',
+        key: 'suggestionSearch' + i,
+        onMouseDown: e => onOpen(href)
+      }, text));
+    }
+
+    for (let i = 0; i < historyCount; i++) {
+      let href = history[i].get('href');
+      let text = history[i].get('text');
+      suggestions.push(DOM.p({
+        className: 'suggestion history',
+        key: 'suggestionHistory' + i,
+        onMouseDown: e => onOpen(href)
+      }, text));
+    }
+
+    return DOM.div({
+      className: 'suggestionscontainer',
+      key: 'suggestionscontainer',
+      style: {
+        backgroundColor: theme.tabstrip.backgroundColor,
+        color: theme.titleText.color
+      }
+    }, [
+      DOM.div({
+        className: 'suggestions',
+        key: 'suggestions'
+      }, suggestions)
+    ])
+  });
+
+
+  let xhrSearch;
+  let xhrHistory;
+
+  const computeSuggestions = (textInput, suggestionsCursor) => {
+
+    if (xhrSearch) xhrSearch.abort();
+    if (xhrHistory) xhrHistory.abort();
+
+    textInput = textInput.trim();
+
+    if (!textInput) {
```

Comment:
Would be better do `suggestionsCursor.merge({search: List(), history: List()})` as that would not trigger two updates.